### PR TITLE
feat: eip681 TransactionRequest from parts

### PR DESCRIPTION
**Pull request by schell**
These changes add constructors for `TransactionRequest` from parts. This will be useful for FFI, when we want to construct a request address from pieces previously parsed.

### CODE REVIEW

File: components/eip681/src/request.rs [hunk 1]
```diff
@@ -52,6 +52,43 @@ pub enum TransactionRequest {
 }
 
 impl TransactionRequest {
+    /// Construct a `TransactionRequest` from `NativeRequest` parts, if possible.
+    pub fn from_native_request_parts(
+        schema_prefix: &str,
```
Comment (schell) [inline components/eip681/src/request.rs:57]: This forgets the possibility of the "pay-" portion of a schema prefix. I'm still thinking of a good way to represent that. I might just have to pull in all of `SchemaPrefix` here.

File: components/eip681/src/request.rs [hunk 2]
```diff
@@ -52,6 +52,47 @@ pub enum TransactionRequest {
...
+    ) -> Result<Self, Error> {
+        let chain_id = chain_id.map(|id| format!("@{id}")).unwrap_or_default();
+        let value = value.map(|v| format!("value={v}")).unwrap_or_default();
+        let gas_limit = gas_limit
+            .map(|v| format!("gasLimit={v}"))
+            .unwrap_or_default();
+        let gas_price = gas_price
+            .map(|v| format!("gasPrice={v}"))
+            .unwrap_or_default();
+        let params = [value, gas_limit, gas_price].join("&");
```
Comment (nuttycom) [inline components/eip681/src/request.rs]: bug: This will result in a sequence of `&` characters if gas limit, gas price, or values are None in the inputs because they get mapped to the empty string, and then concatenated. The `unwrap_or_default`s here are incorrect.

**Review by nuttycom — CHANGES_REQUESTED** [review]
One bug, plus Clippy failures.

File: components/eip681/src/request.rs [hunk 3]
```diff
@@ -343,6 +389,16 @@ impl TryFrom<&RawTransactionRequest> for Erc20Request {
 }
 
 impl Erc20Request {
+    /// Retuns the schema prefix.
```
Comment (nullcopy) [inline components/eip681/src/request.rs]:
```suggestion
    /// Returns the schema prefix.
```

File: components/eip681/src/request.rs [hunk 4]
```diff
@@ -52,6 +52,47 @@ pub enum TransactionRequest {
...
+        let gas_price = gas_price
+            .map(|v| format!("gasPrice={v}"))
+            .unwrap_or_default();
+        let params = [value, gas_limit, gas_price].join("&");
+        let pay = if has_pay { "pay-" } else { "" };
+        let req = format!("{schema_prefix}:{pay}{recipient}{chain_id}?{params}",);
+        Self::parse(&req)
+    }
+
+    // Construct a `TransactionRequest` from `Erc20Request` parts, if possible.
```
Comment (nullcopy) [inline components/eip681/src/request.rs]:
```suggestion
    /// Construct a `TransactionRequest` from `Erc20Request` parts, if possible.
```

File: components/eip681/src/request.rs [hunk 5]
```diff
@@ -452,4 +508,122 @@ mod test {
...
+        assert_eq!(
+            erc20.token_contract_address(),
+            erc20_b.token_contract_address()
+        );
+        assert_eq!(erc20.recipient_address(), erc20_b.recipient_address());
+        assert_eq!(erc20.value_atomic(), erc20_b.value_atomic());
+    }
+
+    #[test]
+    fn from_erc20_request_with_pay_parts_roundtrip() {
```
Comment (nullcopy) [inline components/eip681/src/request.rs]: FYI, none of these tests cover cases where any of the optional parameters are `None`. Might be nice to have at least one test covering edge cases relating to optional arguments

File: components/eip681/src/request.rs (hunk 2 above)
Comment (schell) [reply components/eip681/src/request.rs]: Ah, yeah, good catch.

File: components/eip681/src/request.rs (hunk 5 above)
Comment (schell) [reply components/eip681/src/request.rs]: I think I'll convert these into one proptest. Thanks :)

**Review by nullcopy — APPROVED** [review]
utACK df6b7c498a727802cae929793f5804063635f2f5

File: components/eip681/src/parse.rs [hunk 6]
```diff
@@ -1151,7 +1149,8 @@ pub struct SchemaPrefix {
 
 impl core::fmt::Display for SchemaPrefix {
     fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
-        f.write_str("ethereum:")?;
+        f.write_str(&self.prefix)?;
```
Comment (str4d) [inline components/eip681/src/parse.rs:1154]: This is a fix for a bug in #2134 that was not caught in review.

In general, please separate bugfixes from feature additions (same PR is fine if the bugfix is needed for the PR, as long as they are in separate commits), so it is clear that it was an intentional bugfix (and so the bugfix can be cherry-picked in the event that the prior PR had already been released, in case the feature being added herein were a breaking change that couldn't also be cherry-picked).

File: components/eip681/src/request.rs [hunk 7]
```diff
@@ -459,4 +513,125 @@ mod test {
...
+        "0x5aAeb6053F3E94C9b9A09f33669435E7Ef1BeAed",
+        "0xdbF03B407c01E7cD3CBea99509d93f8DDDC8C6FB",
+    ];
+
+    fn arb_erc55_address() -> impl Strategy<Value = &'static str> {
+        prop::sample::select(ERC55_ADDRESSES)
+    }
+
+    fn arb_schema_prefix() -> impl Strategy<Value = &'static str> {
+        prop::sample::select(&["ethereum", "zcash", "chaos_emerald"][..])
```
Comment (str4d) [inline components/eip681/src/request.rs]: Do not use `zcash` here. EIP 681 is incompatible with ZIP 321, and we should not imply otherwise here.
```suggestion
        prop::sample::select(&["ethereum", "chaos_emerald"][..])
```

File: components/eip681/src/request.rs [hunk 8]
```diff
@@ -459,4 +513,125 @@ mod test {
...
+    }
+
+    fn arb_u256() -> impl Strategy<Value = U256> {
+        any::<u64>().prop_map(U256::from)
```
Comment (str4d) [inline components/eip681/src/request.rs]: This doesn't cover the full `U256` range. The most useful way to fix this (i.e. the way that maps closest to how `proptest` will do its reductions) would be to sample two `u128`s, construct two `U256`s, and then return `(upper << 128) | lower`.

**Review by str4d — CHANGES_REQUESTED** [review]
Reviewed df6b7c498a727802cae929793f5804063635f2f5. utACK after test fixes.

File: components/eip681/src/parse.rs (hunk 6 above)
Comment (schell) [reply components/eip681/src/parse.rs:1154]: I've separated the commits now.

File: components/eip681/src/parse.rs [hunk 9]
```diff
@@ -1048,7 +1048,7 @@ impl Parameters {
         let mut values = BTreeMap::default();
         for p in self.iter() {
             let (number, n) = match p {
-                Parameter::Value(n) => (n.as_i128()?, n),
+                Parameter::Value(n) => (n.as_uint256()?, n),
```
Comment (str4d) [inline components/eip681/src/parse.rs:1051]: Are these more bugfixes identified after my suggested proptest change? These fixes should also be separated out from the feature commit (the commit message for which mentions the change but does not indicate that it is a bugfix).

In this case, given that the test that found the bug requires the new feature, I'd be fine with adding the feature and test, and then applying the unrelated fix in a follow-up commit (this is functionally equivalent to red-green bugfixing which we also do when we figure out an obscure bug: add a test exercising the bug in a commit that can be verified to fail, then in the next commit fix it).
